libbpf-tools/tcptop.c

```diff
@@ -0,0 +1,420 @@
+{
+	struct info_t *i1 = (struct info_t *)obj1;
+	struct info_t *i2 = (struct info_t *)obj2;
+
+	if (i1->key.family != i2->key.family)
+		/*
+		 * i1 - i2 because we want to sort by increasing order (first AF_INET then
+		 * AF_INET6).
+		 */
+		return i1->key.family - i2->key.family;
+
+	if (sort_by == SENT)
+		return i2->value.sent - i1->value.sent;
+	else if (sort_by == RECEIVED)
+		return i2->value.received - i1->value.received;
+	else
+		return (i2->value.sent + i2->value.received) - (i1->value.sent + i1->value.received);
+}
+
+static int print_stat(struct tcptop_bpf *obj)
+{
+	FILE *f;
+	time_t t;
+	struct tm *tm;
+	char ts[16], buf[256];
+	struct ip_key_t key, *prev_key = NULL;
+	static struct info_t infos[OUTPUT_ROWS_LIMIT];
+	int n, i, err = 0;
+	int fd = bpf_map__fd(obj->maps.ip_map);
+	int rows = 0;
+	bool ipv6_header_printed = false;
+
+	if (!no_summary) {
+		f = fopen("/proc/loadavg", "r");
+		if (f) {
+			time(&t);
+			tm = localtime(&t);
+			strftime(ts, sizeof(ts), "%H:%M:%S", tm);
+			memset(buf, 0, sizeof(buf));
+			n = fread(buf, 1, sizeof(buf), f);
+			if (n)
+				printf("%8s loadavg: %s\n", ts, buf);
+			fclose(f);
+		}
+	}
+
+	while (1) {
+		err = bpf_map_get_next_key(fd, prev_key, &infos[rows].key);
+		if (err) {
+			if (errno == ENOENT) {
+				err = 0;
+				break;
+			}
+			warn("bpf_map_get_next_key failed: %s\n", strerror(errno));
+			return err;
+		}
+		err = bpf_map_lookup_elem(fd, &infos[rows].key, &infos[rows].value);
+		if (err) {
+			warn("bpf_map_lookup_elem failed: %s\n", strerror(errno));
+			return err;
+		}
+		prev_key = &infos[rows].key;
+		rows++;
+	}
+
+	printf("%-6s %-12s %-21s %-21s %6s %6s", "PID", "COMM", "LADDR", "RADDR",
+				 "RX_KB", "TX_KB\n");
+
+	qsort(infos, rows, sizeof(struct info_t), sort_column);
+	rows = rows < output_rows ? rows : output_rows;
+	for (i = 0; i < rows; i++) {
+		struct ip_key_t *key = &infos[i].key;
+		struct traffic_t *value = &infos[i].value;
+
+		if (key->family == AF_INET6 && !ipv6_header_printed) {
+			printf("\n%-6s %-12s %-51s %-51s %6s %6s", "PID", "COMM", "LADDR6",
+						 "RADDR6", "RX_KB", "TX_KB\n");
+			ipv6_header_printed = true;
+		}
+
+		/*
+			* A port is stored in u16, so highest value is 65535, which is 5
+			* characters long.
+			* We need one character more for ':'.
+			*/
+		size_t size = INET6_ADDRSTRLEN + PORT_LENGTH + 1;
+
+		char saddr[INET6_ADDRSTRLEN];
+		char daddr[INET6_ADDRSTRLEN];
+		char saddr_port[size];
+		char daddr_port[size];
+
+		inet_ntop(key->family, &key->saddr, saddr, INET6_ADDRSTRLEN);
+		inet_ntop(key->family, &key->daddr, daddr, INET6_ADDRSTRLEN);
+
+		snprintf(saddr_port, size, "%s:%d", saddr, key->lport);
+		snprintf(daddr_port, size, "%s:%d", daddr, key->dport);
+
+		printf(key->family == AF_INET
+					? "%-6d %-12.12s %-21s %-21s %6ld %6ld\n"
+					: "%-6d %-12.12s %-51s %-51s %6ld %6ld\n",
+					key->pid, key->name, saddr_port, daddr_port,
+					value->received / 1024, value->sent / 1024);
```

Comment:
FYI, we have a elegant way to handle this:

https://github.com/iovisor/bcc/blob/c54336ef2b26c7a244604efb8741a5c5b3f0b968/libbpf-tools/tcplife.c#L116-L136